File: apps/mobile/modules/mocomo-native-feed/android/src/main/java/net/mocomo/nativefeed/VideoPoster.kt
package net.mocomo.nativefeed

/**
 * Resolve a still thumbnail for feed video cells.
 * Prefer server posterUrl / streamUid; otherwise derive Cloudflare Stream
 * first-frame URL only from known Stream hosts (never guess from arbitrary hex paths).
 */
object VideoPoster {
  private val CF_UID = Regex(
    """(?:videodelivery\.net|cloudflarestream\.com)/([^/?#]+)""",
    RegexOption.IGNORE_CASE,
  )

  fun resolve(
    posterUrl: String?,
    videoUrl: String?,
    hlsUrl: String? = null,
    streamUid: String? = null,
  ): String? {
    posterUrl?.trim()?.takeIf { it.isNotEmpty() }?.let { return it }

    streamUid?.trim()?.takeIf { it.isNotEmpty() && isPlausibleStreamUid(it) }?.let { uid ->
      return cfThumb(uid)
    }

    val probe = listOfNotNull(hlsUrl, videoUrl)
      .map { it.trim() }
      .firstOrNull { it.isNotEmpty() }
      .orEmpty()
    if (probe.isEmpty()) return null

    val uid = CF_UID.find(probe)?.groupValues?.getOrNull(1) ?: return null
    if (!isPlausibleStreamUid(uid)) return null
    return cfThumb(uid)
  }

  private fun cfThumb(uid: String): String =
    "https://videodelivery.net/$uid/thumbnails/thumbnail.jpg?time=0s&height=720"

  private fun isPlausibleStreamUid(uid: String): Boolean =
    uid.length >= 16 && uid.all { it.isLetterOrDigit() || it == '_' || it == '-' }
}

File: apps/mobile/src/lib/video-poster.ts
import type { FeedMedia } from "@/api/feed";

/** Same poster URL the feed already cached, so STAR cells do not re-download it. */
export function resolveVideoPoster(media: FeedMedia): string | null {
  const direct = media.posterUrl?.trim();
  if (direct) return direct;

  const streamUid = media.streamUid?.trim();
  if (streamUid && /^[a-zA-Z0-9_-]{16,}$/.test(streamUid)) {
    return `https://videodelivery.net/${streamUid}/thumbnails/thumbnail.jpg?time=0s&height=720`;
  }

  // Only derive from known Cloudflare Stream hosts — never guess from arbitrary hex paths
  // (R2/CDN keys often contain 32-hex segments and would 404 on videodelivery.net).
  const probe = media.hlsUrl?.trim() || media.url?.trim() || "";
  const uid =
    probe.match(/videodelivery\.net\/([^/?#]+)/i)?.[1] ||
    probe.match(/cloudflarestream\.com\/([^/?#]+)/i)?.[1];
  if (uid && /^[a-zA-Z0-9_-]{16,}$/.test(uid)) {
    return `https://videodelivery.net/${uid}/thumbnails/thumbnail.jpg?time=0s&height=720`;
  }
  return null;
}
